src/expr-to-mql.js

```diff
@@ -0,0 +1,83 @@
+import jsep from 'jsep';
+import validateAST from './utils/validate-ast';
+import { isNumber } from 'lodash';
+
+// maps binary expression operators to their respective agg operators
+const BINARY_OPERATOR_MAP = {
+  '+': '$sum',
+  '-': '$subtract',
+  '*': '$multiply',
+  '/': '$divide'
+};
+
+// expression builder function for BinaryExpression nodes
+const binaryExpressionFn = node => ({
+  [BINARY_OPERATOR_MAP[node.operator]]: [
+    mapNodeToMQL(node.left),
+    mapNodeToMQL(node.right)
+  ]
+});
+
+// expression builder function for UnaryExpression nodes
+const unaryExpressionFn = node => {
+  // if positive, just continue with argument
+  if (node.operator === '+') {
+    return mapNodeToMQL(node.argument);
+  }
+  // for literals, we can return the negative value
+  if (node.argument.type === 'Literal' && isNumber(node.argument.value)) {
+    return -node.argument.value;
+  }
+  // flatten nested unary expressions
+  if (node.argument.type === 'UnaryExpression') {
+    node.argument.operator = node.argument.operator === '+' ? '-' : '+';
+    return mapNodeToMQL(node.argument);
+  }
+  // for everything else, multiply with -1 in the agg framework
+  return { $multiply: [-1, mapNodeToMQL(node.argument)] };
+};
+
+// expression builder function for Literal nodes
+const literalExpressionFn = node =>
+  isNumber(node.value) ? node.value : `$${node.value}`;
+
+// expression builder function for Identifier nodes
+const identifierExpressionFn = node => `$${node.name}`;
+
+// maps AST node types to their expression builder functions
+const expressionMap = {
+  BinaryExpression: binaryExpressionFn,
+  UnaryExpression: unaryExpressionFn,
+  Literal: literalExpressionFn,
+  Identifier: identifierExpressionFn
+};
+
+/**
+ * recursive function calling different helper functions
+ * depending on the node type.
+ *
+ * @param  {Object} node    AST node object
+ * @return {Object}         MQL representation of the node
+ */
+const mapNodeToMQL = node => expressionMap[node.type](node);
+
+/**
+ * main function to convert an expression string into
+ * an MQL aggregation language object.
+ *
+ * @param  {String} expr    arithmetic expression
+ * @return {Object}         MQL query object
+ */
+function exprToMQL(expr) {
+  // parse expression into AST
+  const ast = jsep(expr);
```

Comment:
By default `jsep` includes a bunch of [unary & binary operations which this code doesn't handle](https://github.com/soney/jsep/blob/8af6eb7a5ee63133f1b45a60006f1ae2aacdd52a/src/jsep.js#L55):

### Unary Operations
`['-', '!', '~', '+']`

### Binary Operations
```
[ '||', '&&', '|', '^', '&', '==', '!=', '===', '!==', '<',  '>',  '<=',  '>=', '<<',  '>>', '>>>', '+', '-', '*', '/', '%']

I strongly recommend that we run `removeAllUnaryOps` and `removeAllBinaryOps` - then add back only the operations that we support.
```

Reply:
This is currently all checked in `validate-ast.js`, for example the expression 'foo || bar' throws an error `Error: Invalid node type "LogicalExpression". Supported node types are: BinaryExpression, UnaryExpression, MemberExpression, Identifier, Literal.` The expression `!foo` throws an error `Error: Invalid unary expression operator "!". Supported unary expression operators are: +, -.`. 

There are also tests for invalid operators in `validate-ast.spec.js`.
I've whitelisted the types of nodes and the unary and binary operators that we want to allow.

I could additionally remove the unary and binary operators we don't want to allow from the parser directly, but the error message we get back is less obvious, it just says `Error: Unexpected "!" at character 0`.